hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -325,22 +345,75 @@ public boolean saveToSnapshot() {
                 }
             }
 
-            Stream<Entry<?, ?>> leftBufferStream = buffer[0].content()
-                    .stream()
-                    .map(row -> entry(
-                            ObjectArrayKey.project(row, joinInfo.leftEquiJoinIndices()),
-                            new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 0)
-                    ));
+            if (joinInfo.isEquiJoin()) {
+                leftBufferStream = buffer[0].content()
+                        .stream()
+                        .map(row -> entry(
+                                ObjectArrayKey.project(row, joinInfo.leftEquiJoinIndices()),
+                                new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 0)
+                        ));
+
+                rightBufferStream = buffer[1].content()
+                        .stream()
+                        .map(row -> entry(
+                                ObjectArrayKey.project(row, joinInfo.rightEquiJoinIndices()),
+                                new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 1)
+                        ));
 
-            Stream<Entry<?, ?>> rightBufferStream = buffer[1].content()
-                    .stream()
-                    .map(row -> entry(
-                            ObjectArrayKey.project(row, joinInfo.rightEquiJoinIndices()),
-                            new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 1)
-                    ));
+            } else {
+                final Iterator<Integer>[] it = new Iterator[]{outerJoinKeys.iterator()};
+
+                // Note : details are described in TDD : docs/design/sql/15-stream-to-stream-join.md
+                if (joinInfo.isRightOuter()) {
+                    if (processorIndex == 0) {
+                        leftBufferStream = mapWithIndex(
+                                buffer[0].content().stream(),
+                                (row, index) -> entry(
+                                        broadcastKey(new BufferSnapshotKey(index)),
+                                        new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 0)
+                                ));
+                    } else {
+                        leftBufferStream = Stream.empty();
+                    }
+                    rightBufferStream = buffer[1].content()
+                            .stream()
+                            .map(row -> {
+                                        if (!it[0].hasNext()) {
```

Comment:
what will happen in `AT_LEAST_ONCE` mode when unicast ("inner" side) source repeats some items after restore from snapshot (it is allowed to do so) and the elements will be sent to different partition/instance than on the first time? The partitioning is not explicit in this case. I think that it is possible in such case that 2 instances can have the same inner-side row: one from the snapshot, other one repeated from source. They both will produce null-padded row or duplicated joined rows which is incorrect.

Reply:
AFAIK, for `AT_LEAST_ONCE` mode it is acceptable.

Reply:
hmm, actually this may be true :) But the behaviour will be counter-intuitive. We can get duplicated rows for some time after restore and even for items that were produced by outer side after restore. This will happen until inner-side rows expire due to watermarks (which may be minutes, days or more).

Reply:
@viliam-durina do you think that such behaviour is acceptable?

Reply:
we can also produce incorrect results for non-equi outer join in AT_LEAST_ONCE mode. Consider such scenario:

1. inner side row is received, no matching outer-side row
2. snapshot is taken
3. job is restored from snapshot
4. inner side row is redelivered and ends up in different processor that the one that was restored from snapshot (*)
5. some matching outer-side rows are received
6. watermark causes row to expire
7. null-padded row is emitted by the processor instance that got inner-side row from snapshot - wrong result

(*) can happen. Eg. if StreamJmsP fails when commiting txn in 2nd phase (eg. connection to broker was lost), and the job is subsequently restarted (not by snapshot failure, but by other error - eg. read error from broker).

Reply:
the underlying cause is that `unicast().local()` does not guarantee predictable partitioning (I had the same problem in streaming nested-loops join implementation). It will be partitioned in any way that the source is partitioned, but we do not know it and cannot (?) control it. In NL-join I decided to repartition by something deterministic but it is an overhead.

Reply:
the example should be other way around: we should receive matching outer-side rows only before snapshot and they are not redelivered (possible, they can come from different source/broker), and wrong null-padded result is produced by the processor instance that got redelivered inner-side row.

Reply:
1. inner side row is received, no matching outer-side row
2. some matching outer-side rows are received
3. snapshot is taken
4. job is restored from snapshot
5. inner side row is redelivered and ends up in different processor that the one that was restored from snapshot (*), outer-side rows are not redelivered
6. watermark causes row to expire
7. null-padded row is emitted by the processor instance that got redelivered inner-side row - wrong result

Reply:
I have doubt regarding the latter scenario. Outer-side rows are broadcast in snapshot. Is it possible that they are removed from buffer before matching inner-side row (and before point 3)? Watermarks are tricky, but if so, then we can produce wrong result.

Reply:
this should be possible if inner-side watermark advances so much that outer-side rows can be expired, but at the same time outer-side watermark does not advance.

Reply:
I turns out that with `AT_LEAST_ONCE` guarantee many unexpected events can occur but they are allowed:

1. watermarks can go backwards
2. joined rows can be duplicated for a long time after restore from snapshot
3. null-padded rows can be duplicated
4. for outer join, both null-padded and joined rows can be emitted
5. duplication can occur for some joined rows but not for others

These events should not occur unless the job is restarted from snapshot.

We could add some warning to the TDD